Indicators/ChaikinMoneyFlow.cs

```diff
@@ -0,0 +1,107 @@
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    /// The Chaikin Money Flow Index (CMF) is a volume-weighted average of accumulation and distribution over
+    /// a specified period.
+    ///
+    /// CMF = n-day Sum of [(((C - L) - (H - C)) / (H - L)) x Vol] / n-day Sum of Vol
+    ///
+    /// Where:
+    /// n = number of periods, typically 21
+    /// H = high
+    /// L = low
+    /// C = close
+    /// Vol = volume
+    /// </summary>
+    public class ChaikinMoneyFlow : WindowIndicator<TradeBar>, IIndicatorWarmUpPeriodProvider
+    {
+        /// <summary>
+        /// Holds the point-wise flow and volume terms. 
+        /// </summary>
+        private readonly RollingWindow<List<decimal>> _rollingFlow;
+
+
+        /// <summary>
+        /// Initializes a new instance of the ChaikinMoneyFlow class
+        /// </summary>
+        /// <param name="name">A name for the indicator</param>
+        /// <param name="period">The period over which to perform computation</param>
+        public ChaikinMoneyFlow(string name, int period)
+            : base($"CMF({name})", period)
+        {
+            WarmUpPeriod = period + 1;
+            _rollingFlow = new RollingWindow<List<decimal>>(period);
+        }
+
+
+        /// <summary>
+        /// Gets a flag indicating when this indicator is ready and fully initialized
+        /// </summary>
+        public override bool IsReady => Samples > Period;
+
+        /// <summary>
+        /// Required period, in data points, for the indicator to be ready and fully initialized.
+        /// </summary>
+        public int WarmUpPeriod { get; }
+
+        /// <summary>
+        /// Resets this indicator to its initial state
+        /// </summary>
+        public override void Reset()
+        {
+            _rollingFlow.Reset();
+            base.Reset();
+        }
+
+        /// <summary>
+        /// Computes the next value for this indicator from the given state.
+        /// </summary>
+        /// <param name="window">The window of data held in this indicator</param>
+        /// <param name="input">The input value to this indicator on this time step</param>
+        /// <returns>A new value for this indicator</returns>
+        protected override decimal ComputeNextValue(IReadOnlyWindow<TradeBar> window, TradeBar input)
+        {
+            var denominator = (input.High - input.Low);
+            var flowRatio = denominator > 0
+                ? input.Volume * (input.Close - input.Low - (input.High - input.Close)) / denominator : 0m;
+            var inputVol = input.Volume;
+            var addFlow = new List<decimal> {flowRatio, inputVol};
+            _rollingFlow.Add(addFlow);
+
+            if (IsReady)
+            {
+                var newFlow = 0m;
+                var newVol = 0m;
+
+                foreach (var flowTerm in _rollingFlow)
+                {
+                    newFlow += flowTerm[0];
+                    newVol += flowTerm[1];
```

Comment:
Generated by Trading View :) I checked with him too!

Reply:
Hello @AlexCatarino , I was just trying to implement your suggestions but have found that using the sum indicator as follows

```c#

// Under initialization 
        public ChaikinMoneyFlow(string name, int period)
            : base($"CMF({name})")
        {
            Period = period;
            WarmUpPeriod = period + 1;
            _rollingFlow = new RollingWindow<List<decimal>>(period);
        }
  
              
// under ComputeNextValue
            var _flowRatioSum = new Sum(Period);
            var _volumeSum = new Sum(Period);
            _flowRatioSum.Update(input.EndTime, flowRatio);
            _volumeSum.Update(input.EndTime, input.Volume);
            return !IsReady || _volumeSum == 0m ? 0m : _flowRatioSum / _volumeSum;
```
will result in the test failing (and misses the target by a lot). 

On the other hand, even after making the other changes with respect changing the type to `TradeBarIndicator`, the old loop method works. Do you have any idea as to why this might be?

Reply:
`_flowRatioSum` and `_volumeSum` are class variables initialized in the constructor. I have tested it and it works as intended.

Reply:
That works!